### ISSUE TITLE
Dreamer V3 Performance

### ISSUE THREAD

**Issue by michele-milesi** [Eclectic-Sheep/sheeprl#218]
Hi @LYK-love, I am bringing your question on the performance of Dreamer v3 back here so that we can continue the conversation.

> I'm worried about the performance of your implementation of DreamerV3 is not as well as the original one (by Hafner). Can you show some evaluation scores of DreamerV3?

**Comment by michele-milesi** [comment]
@LYK-love, I will show you three experiments that we compared with the results described in the Dreamer V3 paper (https://arxiv.org/abs/2301.04104).

1. **Crafter**
<img width="671" alt="crafter-reward" src="https://github.com/Eclectic-Sheep/sheeprl/assets/74559684/5a150cd1-76bd-4c08-8e86-09f64e48e0b1">

The reward we obtained in crafter with [these configs](https://github.com/Eclectic-Sheep/sheeprl/blob/main/sheeprl/configs/exp/dreamer_v3_XL_crafter.yaml). The paper claims to have achieved a reward of `11.7 ± 1.9` which is in line with our results (`12.1` of reward during the test). In addition, our plot of the reward during training is almost equal to the one obtained by Hafner.

2. **MsPacman 100K**
<img width="674" alt="ms-pacman-reward" src="https://github.com/Eclectic-Sheep/sheeprl/assets/74559684/35dd26a6-a0b2-4551-bc30-4b06cf9da851">

We used [these configs](https://github.com/Eclectic-Sheep/sheeprl/blob/main/sheeprl/configs/exp/dreamer_v3_100k_ms_pacman.yaml) for training (+ `fabric.accelerator=cuda`).
The paper obtained a score of `1327`, we evaluated the trained agent with 6 seeds and we obtained the following results (`1.911 ± 505.78`):
```txt
MsPacman Test Reward
2020.0 (seed 5)
1070.0 (seed 1024)
2050.0 (seed 42)
1940.0 (seed 1337)
2630.0 (seed 8)
1760.0 (seed 2)
```

3. **Boxing 100K**
<img width="676" alt="boxing-reward" src="https://github.com/Eclectic-Sheep/sheeprl/assets/74559684/8de7d8de-41d6-4088-8c97-9b793254128a">

We used [these configs](https://github.com/Eclectic-Sheep/sheeprl/blob/main/sheeprl/configs/exp/dreamer_v3_100k_boxing.yaml) for training (+ `fabric.accelerator=cuda`).
The paper obtained a score of `78`, we evaluated the trained agent with 6 seeds and we obtained the following results (`94 ± 2.53`):
```txt
Boxing Test Reward
96.0 (seed 5)
92.0 (seed 1024)
96.0 (seed 42)
90.0 (seed 1337)
94.0 (seed 8)
96.0 (seed 2)
```

Let me know if you have other questions regarding the performance of Dreamer V3.
Thanks

**Comment by LYK-love** [comment]
How can I use these configs? I tried to train Crafter with `python sheeprl.py exp=dreamer_v3_XL_crafter env=crafter env.id=crafter fabric.accelerator=gpu fabric.strategy=ddp fabric.devices=8` (I have 8 gpus) but got error:
```text
Error executing job with overrides: ['exp=dreamer_v3_XL_crafter', 'fabric.accelerator=gpu', 'fabric.strategy=ddp', 'fabric.devices=8']
Error locating target 'sheeprl.envs.crafter.CrafterWrapper', set env var HYDRA_FULL_ERROR=1 to see chained exception.
```
Thanks!

**Comment by michele-milesi** [comment]
Hi @LYK-love, we have never tried those two environments.

**Comment by LYK-love** [comment]
Hello, I got the rewards, but I think I made some mistake when training. 
# Crafter reward
This is my training reward for `crafter_reward`. It only runs for 215,000 steps, instead of 1, 000, 000 steps in your [config](https://github.com/Eclectic-Sheep/sheeprl/blob/main/sheeprl/configs/exp/dreamer_v3_XL_crafter.yaml). I don't know why.
<img width="966" alt="image" src="https://github.com/Eclectic-Sheep/sheeprl/assets/74082944/54fd1f14-f3b5-4330-bcb9-138e4710c97a">
The reward value is lower--only `5.1` compared to your `12.1`.

<img width="867" alt="image" src="https://github.com/Eclectic-Sheep/sheeprl/assets/74082944/f69caf77-a2cf-41a2-a5e3-d54c09a143d6">


I evaluated this trained agent with checkpoint `200000.ckpt`:
```bash
export CKPT="logs/runs/dreamer_v3/crafter_reward/2024-03-15_02-26-07_dreamer_v3_crafter_reward_5/version_0/checkpoint/ckpt_200000_0.ckpt"
python sheeprl_eval.py checkpoint_path=$CKPT fabric.accelerator=gpu env.capture_video=True
```
I got evaluation reward `5.09`.

# MsPacman 100K


This is my reward for Pacman.  It runs for 100, 000 steps which is the same as the number in your [config](https://github.com/Eclectic-Sheep/sheeprl/blob/main/sheeprl/configs/exp/dreamer_v3_100k_ms_pacman.yaml). 
<img width="964" alt="image" src="https://github.com/Eclectic-Sheep/sheeprl/assets/74082944/16d35baf-34f6-4230-80a4-3462866914ab">

However, the reward value is also lower--only `570` compared to your `1327`.
<img width="971" alt="image" src="https://github.com/Eclectic-Sheep/sheeprl/assets/74082944/f2eb865d-93df-4b64-8e63-28c4a3dd729e">

Meanwhile, at step = 90, 000. I do observed a reward=1300, which is similar to `1327`.
<img width="973" alt="image" src="https://github.com/Eclectic-Sheep/sheeprl/assets/74082944/13df7fe3-f9f0-4bbf-9bb1-42c8c3b08f6d">

I evaluated this trained agent with checkpoint at 100,000 steps, and set 6 seeds. The commands are
```bash
export CKPT="logs/runs/dreamer_v3/MsPacmanNoFrameskip-v4/2024-03-15_02-20-34_dreamer_v3_MsPacmanNoFrameskip-v4_5/version_0/checkpoint/ckpt_100000_0.ckpt"

seeds=(5 1024 42 1337 8 2)

for seed in "${seeds[@]}"; do
  python sheeprl_eval.py checkpoint_path=$CKPT fabric.accelerator=gpu env.capture_video=True seed=$seed
done
```

The evaluation rewards are
```
Test - Reward: 1730.0 (seed=5)
Test - Reward: 570.0 (seed=1024)
Test - Reward: 810.0 (seed=42)
Test - Reward: 640.0  (seed=1337)
Test - Reward: 540.0 (seed=8)
Test - Reward: 580.0(seed=2)
```
The average evaluation reward is `811.67`


# Boxing

This is my training reward for Boxing.  The training has 100, 000 steps which is the same as the number in your [config](https://github.com/Eclectic-Sheep/sheeprl/blob/main/sheeprl/configs/exp/dreamer_v3_100k_boxing.yaml).
<img width="989" alt="image" src="https://github.com/Eclectic-Sheep/sheeprl/assets/74082944/d737ac59-c3d1-4071-95e4-567b232ba76f">

However, the reward is for 85,000 steps,  instead of 100, 000 steps. I don't know why. The reward value is lower as well--only `18`, instead of your `94 ± 2.53`.
<img width="976" alt="image" src="https://github.com/Eclectic-Sheep/sheeprl/assets/74082944/64487183-c4be-45d4-9fa6-dee361a37d85">
I evaluated this trained agent with checkpoint at 100,000 steps, and set 6 seeds. The commands are
```bash
export CKPT="logs/runs/dreamer_v3/BoxingNoFrameskip-v4/2024-03-15_02-28-28_dreamer_v3_BoxingNoFrameskip-v4_5/version_0/checkpoint/ckpt_100000_0.ckpt"

seeds=(5 1024 42 1337 8 2)

for seed in "${seeds[@]}"; do
  python sheeprl_eval.py checkpoint_path=$CKPT fabric.accelerator=gpu env.capture_video=True seed=$seed
done
```

The evaluation rewards are
```
Test - Reward: 32.0 (seed=5)
Test - Reward: 437.0 (seed=1024)
Test - Reward: 23.0 (seed=42)
Test - Reward: 28.0  (seed=1337)
Test - Reward: 32.0 (seed=8)
Test - Reward: 44.0 (seed=2)
```
The average evaluation reward is `29.33`

# Conclusion
I have two questions:

1.  In `crafter_reward`, the only runs for 215,000 steps, instead of 1, 000, 000 steps. Why?
2. All the final training rewards and evaluation rewards are lower than your data. Why?

I used 8 gpus for training, may be this is one reason?

**Comment by michele-milesi** [comment]
Hi @LYK-love,
1. This is strange, if you set 1M of total (policy) steps in the config, then 1M policy steps are performed. As happened in Atari 100K: you used 8 GPUs, but the number of steps is still 100k. So it should not be related to the number of GPUs you used for training.
2. With distributed training some considerations must be made. First, I suggest reading this discussion: https://github.com/Lightning-AI/pytorch-lightning/discussions/3706#discussioncomment-3960433, in which it is suggested to divide the batch size by the number of GPUs.
Another thing we are working on is how to manage how often to update the model (train_every) and how often to update the parameters (per_rank_gradient_steps). In distributed training, these parameters should be modified according to the number of GPUs used. You can find an explanation of how we would like to replace the `train_every` and `per_rank_gradient_steps` in issue #223.
This means that your training is different from ours.

In the meantime, I advise you not to distribute the training, at least not until we fix this.
Sorry for that.

**Comment by LYK-love** [comment]
Sure. I wonder what is the commands for training. When I use
```bash
python sheeprl.py exp=dreamer_v3_XL_crafter fabric.accelerator=cuda
```
I got an error:
```
Seed set to 5
Log dir: logs/runs/dreamer_v3/crafter_reward/2024-03-18_04-11-30_dreamer_v3_crafter_reward_5/version_0
Error executing job with overrides: ['exp=dreamer_v3_XL_crafter']
Error locating target 'sheeprl.envs.crafter.CrafterWrapper', set env var HYDRA_FULL_ERROR=1 to see chained exception.

Set the environment variable HYDRA_FULL_ERROR=1 for a complete stack trace.
```

Meanwhile, I didn't get any error when running:
```bash
python sheeprl.py exp=dreamer_v3_100k_boxing
```

**Comment by michele-milesi** [comment]
Can you share your environment? There may be a problem with the `ruamel.yaml` package: we fixed it with PR no. #230.
Which commit are you using?
I suggest updating the repo and installing Crafter with `pip install -e .[crafter]`.

Let me know, thanks.

**Comment by belerico** [comment]
Hi @LYK-love, this is experiment that I've run on Ms-PacMan: https://github.com/Eclectic-Sheep/sheeprl/issues/261#issuecomment-2047994119.

It has been run with the torch.compie model, but it contains all the improvements @michele-milesi listed here

**Comment by belerico** [comment]
@LYK-love  I'm closing this due to both inactivity and it seems to have been resolved. Re-open it if you have more evidence on your side
